ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionHeartbeatService.java

```diff
@@ -0,0 +1,206 @@
+   * @throws InterruptedException
+   */
+  void shutdown() throws InterruptedException {
+    shuttingDown = true;
+    LOG.info("Shutting down compaction txn heartbeater service.");
+    for (CompactionHeartbeater heartbeater : tasks.values()) {
+      heartbeater.stop();
+    }
+    tasks.clear();
+    clientPool.close();
+    LOG.info("Compaction txn heartbeater service is successfully stopped.");
+  }
+
+  private CompactionHeartbeatService(HiveConf conf) {
+    int numberOfWorkers = MetastoreConf.getIntVar(conf, MetastoreConf.ConfVars.COMPACTOR_WORKER_THREADS);
+    GenericObjectPoolConfig<IMetaStoreClient> config = new GenericObjectPoolConfig<>();
+    config.setMinIdle(1);
+    config.setMaxIdle(2);
+    config.setMaxTotal(numberOfWorkers);
+    config.setBlockWhenExhausted(true);
+    config.setMaxWaitMillis(2000);
+    config.setTestOnBorrow(false);
+    config.setTestOnCreate(false);
+    config.setTestOnReturn(false);
+    config.setTestWhileIdle(false);
+    clientPool = new GenericObjectPool<>(new IMetaStoreClientFactory(conf), config);
+    long txnTimeout = MetastoreConf.getTimeVar(conf, MetastoreConf.ConfVars.TXN_TIMEOUT, TimeUnit.MILLISECONDS);
+    initialDelay = txnTimeout / 4;
+    period = txnTimeout / 2;
+  }
+
+  private final class CompactionHeartbeater {
+    final private Logger LOG = LoggerFactory.getLogger(CompactionHeartbeater.class);
+    private final long txnId;
+    private final long lockId;
+    private final String tableName;
+    private final ScheduledThreadPoolExecutor heartbeatExecutor;
+
+
+    void start() {
+      heartbeatExecutor.scheduleAtFixedRate(() -> {
+        IMetaStoreClient msc = null;
+        try {
+          LOG.debug("Heartbeating compaction transaction id {} for table: {}", txnId, tableName);
+          // Create a metastore client for each thread since it is not thread safe
+          msc = clientPool.borrowObject();
+          msc.heartbeat(txnId, lockId);
+          clientPool.returnObject(msc);
+        } catch (NoSuchElementException e) {
+          LOG.error("Compaction transaction heartbeater pool exhausted, unable to heartbeat", e);
+        } catch (TException e) {
+          LOG.error("Error while heartbeating compaction transaction id {} for table: {}", txnId, tableName, e);
+        } catch (Exception e) {
+          LOG.error("Error while heartbeating compaction transaction id {} for table: {}", txnId, tableName, e);
+          if (msc != null) {
+            try {
+              clientPool.invalidateObject(msc);
+            } catch (Exception ex) {
+              LOG.error("Error while invalidating a broken MetaStoreClient instance", e);
+            }
+          }
+        }
+      }, initialDelay, period, TimeUnit.MILLISECONDS);
+    }
+
+    public void stop() throws InterruptedException {
+      LOG.info("Shutting down compaction txn heartbeater instance.");
+      heartbeatExecutor.shutdownNow();
+      heartbeatExecutor.awaitTermination(5000, TimeUnit.MILLISECONDS);
```

Comment:
try {
          if (!heartbeatExecutor.awaitTermination(5, TimeUnit.SECONDS)) {
            LOG.warn("Heartbeating for transaction {} did not stop in 5 seconds, do not wait any longer.", this);
          }
        } catch (InterruptedException ex) {
          //Caller thread was interrupted while waiting for heartbeater to terminate.
          //Nothing to do, just restore the interrupted state.
          Thread.currentThread().interrupt();
        }